services/src/main/java/org/keycloak/authorization/authorization/AuthorizationTokenService.java

```diff
@@ -402,19 +426,25 @@ private ResourceServer getResourceServer(PermissionTicketToken ticket, KeycloakA
         String issuedFor = ticket.getIssuedFor();
 
         if (issuedFor == null) {
-            throw new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "You must provide the issuedFor", Status.BAD_REQUEST);
+            CorsErrorResponseException missingIssuedForException = new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "You must provide the issuedFor", Status.BAD_REQUEST);
+            fireErrorEvent(request.getEvent(), Errors.MISSING_ISSUED_FOR, missingIssuedForException);
+            throw missingIssuedForException;
         }
 
         ClientModel clientModel = request.getRealm().getClientByClientId(issuedFor);
 
         if (clientModel == null) {
-            throw new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "Unknown resource server id.", Status.BAD_REQUEST);
+            CorsErrorResponseException unknownServerIdException = new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "Unknown resource server id: [" + issuedFor + "]", Status.BAD_REQUEST);
+            fireErrorEvent(request.getEvent(), Errors.UNKNOWN_SERVER_ID, unknownServerIdException);
+            throw unknownServerIdException;
         }
 
         ResourceServer resourceServer = resourceServerStore.findById(clientModel.getId());
 
         if (resourceServer == null) {
-            throw new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "Client does not support permissions", Status.BAD_REQUEST);
+            CorsErrorResponseException unsupportedPermissionsException = new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_REQUEST, "Client does not support permissions", Status.BAD_REQUEST);
+            fireErrorEvent(request.getEvent(), Errors.UNSUPPORTED_PERMISSIONS, unsupportedPermissionsException);
```

Comment:
I would just use Errors.INVALID_REQUEST. The details should be enough to understand what happened.

Reply:
agreed